webapp/src/main/java/com/box/l10n/mojito/service/tm/TMTextUnitVariantRepository.java

```diff
@@ -17,9 +27,130 @@ public interface TMTextUnitVariantRepository extends JpaRepository<TMTextUnitVar
+    Page<TMTextUnitVariant> findAllUsedForRepositoryAndLocalesInDateRange(
+            @Param("repository") Repository repository,
+            @Param("locales") List<Locale> locales,
+            @Param("fromDate") DateTime fromDate,
+            @Param("toDate") DateTime toDate,
+            Pageable pageable);
+
+    /**
+     * Gets all {@link TMTextUnitVariant} entities that are linked to a PullRun,
+     * regardless of used status or the asset being deleted or not.
+     */
+    @Query(value = "select distinct tuv from TMTextUnitVariant tuv " +
+            "inner join PullRunTextUnitVariant prtuv on prtuv.tmTextUnitVariant = tuv " +
+            "inner join TMTextUnitCurrentVariant tucv on tucv.tmTextUnitVariant = tuv " +
+            "inner join TMTextUnit tu on tu = tucv.tmTextUnit " +
+            "inner join Asset a on a = tu.asset " +
+            "where a.repository = :repository " +
+            "and tuv.locale in :locales " +
+            "and prtuv.pullRunAsset.pullRun in :pullRuns ")
+    Page<TMTextUnitVariant> findAllVariantsForPullRuns(
+            @Param("repository") Repository repository,
+            @Param("locales") List<Locale> locales,
+            @Param("pullRuns") List<PullRun> pullRuns,
+            Pageable pageable);
+
+    /**
+     * Gets all {@link TMTextUnitVariant} entities that are linked to a PushRun,
+     * regardless of used status or the asset being deleted or not.
+     */
+    @Query(value = "select distinct tuv from TMTextUnitVariant tuv " +
+            "inner join TMTextUnitCurrentVariant tucv on tucv.tmTextUnitVariant = tuv " +
+            "inner join TMTextUnit tu on tu = tucv.tmTextUnit " +
+            "inner join PushRunAssetTmTextUnit prattu on prattu.tmTextUnit = tu " +
+            "inner join PushRunAsset pra on pra = prattu.pushRunAsset " +
+            "inner join Asset a on a = tu.asset " +
+            "where a.repository = :repository " +
+            "and tuv.locale in :locales " +
+            "and pra.pushRun in :pushRuns")
+    Page<TMTextUnitVariant> findAllVariantsForPushRuns(
+            @Param("repository") Repository repository,
+            @Param("locales") List<Locale> locales,
+            @Param("pushRuns") List<PushRun> pushRuns,
+            Pageable pageable);
 
+    /**
+     * Gets all {@link TMTextUnitVariant} entities for the {@link com.box.l10n.mojito.entity.TMTextUnit}
+     * associated with the provided {@link PushRun} that have different values
+     * than the {@link TMTextUnitVariant} entities from the associated
+     * {@link PullRun}.
+     * <p>
+     * Additional filtering done by {@link Repository} ID and a list of
+     * {@link Locale} IDs.
+     * <p>
+     * To note: unused text units and deleted assets are also included,
+     * as this method is indented to provided deltas relevant for a specific snapshot.
+     */
+    @Query(nativeQuery = true,
+            value = "select distinct base_tu.name as tmTextUnitName, " +
```

Comment:
do you have a sense of the performance impact of that query?

Reply:
Not empirically, though the query is optimized as much as it can be for now with predicates being pushed down for both subqueries, the number of rows returned should be quite small in practice.

Once we have production data from pull on large repos - we can measure and tweak this further.